ledger/participant-integration-api/src/test/suite/scala/platform/store/appendonlydao/events/BufferedTransactionsReaderSpec.scala

```diff
@@ -0,0 +1,205 @@
+          apiResponseCtor = apiResponseCtor,
+          fetchTransactions = fetchTransactions,
+          sourceTimer = metrics.daml.services.index.streamsBuffer.getTransactionTrees,
+          resolvedFromBufferCounter =
+            metrics.daml.services.index.streamsBuffer.transactionTreesBuffered,
+          totalRetrievedCounter = metrics.daml.services.index.streamsBuffer.transactionTreesTotal,
+          bufferSizeCounter = metrics.daml.services.index.transactionTreesBufferSize,
+          outputStreamBufferSize = 128,
+        )
+        .runWith(Sink.seq)
+
+    "request within buffer range inclusive on offset gap as start exclusive" should {
+      "fetch from buffer" in {
+        readerGetTransactionsGeneric(
+          eventsBuffer = transactionsBuffer,
+          startExclusive = predecessor(offset3),
+          endInclusive = offset4,
+          fetchTransactions = (_, _, _, _) => fail("Should not fetch"),
+        ).map(
+          _ should contain theSameElementsInOrderAs Seq(
+            offset3 -> apiResponse3,
+            offset4 -> apiResponse4,
+          )
+        )
+      }
+    }
+
+    "request within buffer range inclusive on existing offset as start inclusive" should {
+      "fetch from buffer" in {
+        readerGetTransactionsGeneric(
+          eventsBuffer = transactionsBuffer,
+          startExclusive = offset2,
+          endInclusive = offset4,
+          fetchTransactions = (_, _, _, _) => fail("Should not fetch"),
+        ).map(
+          _ should contain theSameElementsInOrderAs Seq(
+            offset3 -> apiResponse3,
+            offset4 -> apiResponse4,
+          )
+        )
+      }
+    }
+
+    "request before buffer start" should {
+      "fetch from buffer and storage" in {
+        readerGetTransactionsGeneric(
+          eventsBuffer = transactionsBuffer,
+          startExclusive = offset1,
+          endInclusive = offset3,
+          fetchTransactions = {
+            case (`offset1`, `offset2`, `filter`, false) =>
+              Source.single(offset2 -> apiResponse4)
+            case unexpected => fail(s"Unexpected $unexpected")
+          },
+        ).map(
+          _ should contain theSameElementsInOrderAs Seq(
+            offset2 -> apiResponse4, // apiResponse4 on purpose in order to distinguish from fetched from buffer
+            offset3 -> apiResponse3,
+          )
+        )
+      }
+    }
+
+    "request before buffer bounds" should {
+      "fetch only from storage" in {
+        val transactionsBuffer = new EventsBuffer[Offset, TransactionLogUpdate](
+          maxBufferSize = 1L,
+          metrics = metrics,
+          bufferQualifier = "test",
+          isRangeEndMarker = _.isInstanceOf[TransactionLogUpdate.LedgerEndMarker],
+        )
+
+        offsetUpdates.foreach { case (offset, update) =>
+          transactionsBuffer.push(offset, update)
+        }
+        val fetchedElements = Vector(offset2 -> apiResponse1, offset3 -> apiResponse2)
```

Comment:
I think it would be more readable if the elements fetched from storage were saying "Some API response $idx from storage".

Reply:
Done